arangod/VocBase/Identifiers/RevisionId.cpp

```diff
@@ -0,0 +1,168 @@
+
+#include <cstdint>
+#include <limits>
+
+#include "VocBase/Identifiers/RevisionId.h"
+
+#include "Basics/HybridLogicalClock.h"
+#include "Basics/NumberUtils.h"
+#include "Basics/StaticStrings.h"
+#include "Basics/StringUtils.h"
+#include "Basics/debugging.h"
+#include "Cluster/ClusterFeature.h"
+#include "Cluster/ClusterInfo.h"
+#include "Logger/LogMacros.h"
+#include "RocksDBEngine/RocksDBFormat.h"
+#include "VocBase/Identifiers/LocalDocumentId.h"
+
+namespace {
+/// @brief Convert a revision ID to a string
+constexpr static arangodb::RevisionId::BaseType TickLimit =
+    static_cast<arangodb::RevisionId::BaseType>(2016ULL - 1970ULL) * 1000ULL *
+    60ULL * 60ULL * 24ULL * 365ULL;
+}  // namespace
+
+namespace arangodb {
+
+RevisionId::RevisionId(LocalDocumentId const& id) : Identifier(id.id()) {}
+
+bool RevisionId::isSet() const noexcept { return id() != 0; }
+
+bool RevisionId::empty() const noexcept { return !isSet(); }
+
+// @brief get the next revision id in sequence (this + 1)
+RevisionId RevisionId::next() const { return RevisionId{id() + 1}; }
+
+// @brief get the previous revision id in sequence (this - 1)
+RevisionId RevisionId::previous() const { return RevisionId{id() - 1}; }
+
+std::string RevisionId::toString() const {
+  if (id() <= ::TickLimit) {
+    return arangodb::basics::StringUtils::itoa(id());
+  }
+  return basics::HybridLogicalClock::encodeTimeStamp(id());
+}
+
+/// encodes the uint64_t timestamp into the provided result buffer
+/// the buffer must be at least 11 chars long
+/// the length of the encoded value and the start position into
+/// the result buffer are returned by the function
+std::pair<size_t, size_t> RevisionId::toString(char* buffer) const {
+  if (id() <= ::TickLimit) {
+    std::pair<size_t, size_t> pos{0, 0};
+    pos.second = basics::StringUtils::itoa(id(), buffer);
+    return pos;
+  }
+  return basics::HybridLogicalClock::encodeTimeStamp(id(), buffer);
+}
+
+/// encodes the uint64_t timestamp into a temporary velocypack ValuePair,
+/// using the specific temporary result buffer
+/// the result buffer must be at least 11 chars long
+arangodb::velocypack::ValuePair RevisionId::toValuePair(char* buffer) const {
+  auto positions = toString(buffer);
+  return arangodb::velocypack::ValuePair(&buffer[0] + positions.first, positions.second,
+                                         velocypack::ValueType::String);
+}
+
+/// @brief Write revision ID to string for storage with correct endianness
+void RevisionId::toPersistent(std::string& buffer) const {
+  rocksutils::uint64ToPersistent(buffer, id());
+}
+
+/// @brief create a revision id using an HLC value
+RevisionId RevisionId::create() { return RevisionId{TRI_HybridLogicalClock()}; }
+
+/// @brief create a revision id which is guaranteed to be unique cluster-wide
+RevisionId RevisionId::createClusterWideUnique(application_features::ApplicationServer& server) {
+  ClusterFeature& cf = server.getFeature<ClusterFeature>();
```

Comment:
do we really need that lookup everyhwere ? Can the code not use the ClusterInfo directly